src/NuGet.Clients/PackageManagement.VisualStudio/IDE/NuGetAndEnvDTEProjectCache.cs

```diff
@@ -117,38 +119,37 @@ public bool IsAmbiguous(string shortName)
         }
 
         /// <summary>
-        /// Add a project to the cache.
+        /// Adds a project to the project cache. If the project already exists in the cache, this
+        /// this operation does nothing.
         /// </summary>
-        /// <param name="project">project to add to the cache.</param>
-        /// <returns>The project name of the added project.</returns>
-        public EnvDTEProjectName AddProject(EnvDTEProject project, VSNuGetProjectFactory factory)
+        /// <param name="projectName">The project name.</param>
+        /// <param name="project">The VS project.</param>
+        /// <param name="nuGetProject">The NuGet project.</param>
+        /// <returns>
+        /// Returns true if the project was newly added to the cache. Returns false if the project
+        /// was already in the cache.
+        /// </returns>
+        public bool AddProject(EnvDTEProjectName projectName, EnvDTEProject project, NuGetProject nuGetProject)
         {
-            ThreadHelper.ThrowIfNotOnUIThread();
```

Comment:
Since you removed the thread guard from here should all dictionary based caches in this class be replaced with concurrent dictionaries?